Replace exp(x) - 1 with expm1(x)

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92116
* #92115
* __->__ #92154
* #92114

This offers improved precision near zero where `exp(x)` is `1 + O(x)` and doing
`(1 + O(x)) - 1` will truncate anything below the float epsilon to zero.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

peterbell10 committed:
Replace exp(x) - 1 with expm1(x)

This offers improved precision near zero where `exp(x)` is `1 + O(x)` and doing
`(1 + O(x)) - 1` will truncate anything below the float epsilon to zero.

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace exp(x) - 1 with expm1(x)"


This offers improved precision near zero where `exp(x)` is `1 + O(x)` and doing
`(1 + O(x)) - 1` will truncate anything below the float epsilon to zero.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace exp(x) - 1 with expm1(x)"


This offers improved precision near zero where `exp(x)` is `1 + O(x)` and doing
`(1 + O(x)) - 1` will truncate anything below the float epsilon to zero.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace exp(x) - 1 with expm1(x)"


This offers improved precision near zero where `exp(x)` is `1 + O(x)` and doing
`(1 + O(x)) - 1` will truncate anything below the float epsilon to zero.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace exp(x) - 1 with expm1(x)"


This offers improved precision near zero where `exp(x)` is `1 + O(x)` and doing
`(1 + O(x)) - 1` will truncate anything below the float epsilon to zero.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

lezcano reviewed (approved):
Yup!

lezcano commented on torch/distributions/log_normal.py:
```diff
@@ -54,7 +54,7 @@ def mode(self):
     @property
     def variance(self):
-        return (self.scale.pow(2).exp() - 1) * (2 * self.loc + self.scale.pow(2)).exp()
+        return self.scale.pow(2).expm1() * (2 * self.loc + self.scale.pow(2)).exp()
```
Now that you're at it, could you factor out that `self.scale.pow(2)` to avoid computing it twice?

peterbell10 committed:
Update on "Replace exp(x) - 1 with expm1(x)"


This offers improved precision near zero where `exp(x)` is `1 + O(x)` and doing
`(1 + O(x)) - 1` will truncate anything below the float epsilon to zero.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace exp(x) - 1 with expm1(x)"


This offers improved precision near zero where `exp(x)` is `1 + O(x)` and doing
`(1 + O(x)) - 1` will truncate anything below the float epsilon to zero.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace exp(x) - 1 with expm1(x)"


This offers improved precision near zero where `exp(x)` is `1 + O(x)` and doing
`(1 + O(x)) - 1` will truncate anything below the float epsilon to zero.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Lint on "Replace exp(x) - 1 with expm1(x)"


This offers improved precision near zero where `exp(x)` is `1 + O(x)` and doing
`(1 + O(x)) - 1` will truncate anything below the float epsilon to zero.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace exp(x) - 1 with expm1(x)"


This offers improved precision near zero where `exp(x)` is `1 + O(x)` and doing
`(1 + O(x)) - 1` will truncate anything below the float epsilon to zero.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Replace exp(x) - 1 with expm1(x)"


This offers improved precision near zero where `exp(x)` is `1 + O(x)` and doing
`(1 + O(x)) - 1` will truncate anything below the float epsilon to zero.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
